cpp/src/io/parquet/writer_impl.cu

```diff
@@ -1407,17 +1451,54 @@ void writer::impl::write(table_view const& table, std::vector<partition_info> co
-  // If using the default fragment size, scale it up or down depending on the requested page size.
-  if (max_page_fragment_size_ == cudf::io::default_max_page_fragment_size) {
-    max_page_fragment_size_ = (cudf::io::default_max_page_fragment_size * max_page_size_bytes) /
-                              cudf::io::default_max_page_size_bytes;
+  // 5000 is good enough for up to ~200-character strings. Longer strings and deeply nested columns
+  // will start producing fragments larger than the desired page size, so calculate fragment sizes
+  // for each leaf column.  Skip if the fragment size is not the default.
+  bool is_default_fragment_size = max_page_fragment_size_ == -1;
+  if (is_default_fragment_size) { max_page_fragment_size_ = default_max_page_fragment_size; }
+
+  std::vector<size_type> column_frag_size(num_columns, max_page_fragment_size_);
+
+  if (table.num_rows() > 0 && is_default_fragment_size) {
+    std::vector<size_t> column_sizes;
+    std::transform(single_streams_table.begin(),
+                   single_streams_table.end(),
+                   std::back_inserter(column_sizes),
+                   [this](auto const& column) { return column_size(column, stream); });
+
+    // adjust global fragment size if a single fragment will overrun a rowgroup
+    auto const table_size  = std::reduce(column_sizes.begin(), column_sizes.end());
+    auto const avg_row_len = util::div_rounding_up_safe<size_t>(table_size, table.num_rows());
+    if (avg_row_len > 0) {
+      auto const rg_frag_size = util::div_rounding_up_safe(max_row_group_size, avg_row_len);
+      max_page_fragment_size_ = std::min<size_type>(rg_frag_size, max_page_fragment_size_);
+    }
+
+    // dividing page size by average row length will tend to overshoot the desired
+    // page size when there's high variability in the row lengths. instead, shoot
+    // for multiple fragments per page to smooth things out. using 2 was too
+    // unbalanced in final page sizes, so using 4 which seems to be a good
+    // compromise at smoothing things out without getting fragment sizes too small.
+    auto frag_size_fn = [&](auto const& col, size_type col_size) {
+      const int target_frags_per_page = is_col_fixed_width(col) ? 1 : 4;
+      auto const avg_len =
+        target_frags_per_page * util::div_rounding_up_safe<size_type>(col_size, table.num_rows());
+      if (avg_len > 0) {
+        auto const frag_size = util::div_rounding_up_safe<size_type>(max_page_size_bytes, avg_len);
+        return std::min<size_type>(max_page_fragment_size_, frag_size);
+      } else {
+        return max_page_fragment_size_;
+      }
+    };
+
+    std::transform(single_streams_table.begin(),
+                   single_streams_table.end(),
+                   column_sizes.begin(),
+                   column_frag_size.begin(),
+                   frag_size_fn);
   }
 
+  // do first pass with max_page_fragment_size_ to calculate row group boundaries
```

Comment:
Because at this point it should be either the default 5000, a scaled down value if the rows are very wide, or a user supplied value (which is then always used).

Reply:
To be more precise with my question - why don't we use the per-column fragment sizes here?

Reply:
Because each column chunk in the row group has to have the same number of rows, and ensuring this with variable fragment sizes seems very tricky.  It's much easier to use uniform fragment sizes to get the boundaries of each row group, and then readjust when necessary.

Reply:
This might be a TL;DR, but last night I thought more on this question.  Let's assume a table with 3 columns, each wanting a different fragment size.  As this PR stands, pass 1 first produces rowgroup fragments to get the rowgroup boundaries, and then does a second pass to get the page fragments.
```
           curr PR pass 1                 pass 2
 
       C0      C1      C2           C0      C1      C2
     -----------------------       -----------------------
RG0 | F0    | F0    | F0    | RG0 | F0    | F0    | F0    |
    |       |       |       |     |       |       |-------|
    |       |       |       |     |       |       | F1    |
    |       |       |       |     |       |-------|-------|
    |       |       |       |     |       | F1    | F2    |
    |-------|-------|-------|     |-------|       |-------|
    | F1    | F1    | F1    |     | F1    |       | F3    |
    |       |       |       |     |       |-------|-------|
    |       |       |       |     |       | F2    | F4    |
    |       |       |       |     |       |       |-------|
    |       |       |       |     |       |       | F5    |
    |=======|=======|=======|     |=======|=======|=======|
RG1 | F0    | F0    | F0    | RG1 | F0    | F0    | F0    |
    |       |       |       |     |       |       |-------|
    |       |       |       |     |       |       | F1    |
    |       |       |       |     |       |-------|-------|
    |       |       |       |     |       | F1    | F2    |
    |-------|-------|-------|     |-------|       |-------|
    | F1    | F1    | F1    |     | F1    |       | F3    |
     ------- ------- -------       ------- ------- -------
```
One way to do a single pass would be to calculate the number of fragments for C1 and C2 in mini batches of frag_len(C0).  This keeps the RG boundaries sane, but leads to very different fragment sizes for C1 (we could make C0 a multiple of C2 to help with those two, but if there are many columns each with a different desired size it may not be worth it.
```
       C0      C1      C2
     -----------------------
RG0 | F0    | F0    | F0    | 21 fragments total
    |       |       |-------|
    |       |       | F1    |
    |       |-------|-------|
    |       | F1    | F2    |
    |-------|-------|-------|
    | F1    | F2    | F3    |
    |       |       |-------|
    |       |       | F4    |
    |       |-------|-------|
    |       | F3    | F5    |
    |=======|=======|=======|
RG1 | F0    | F0    | F0    |
    |       |       |-------|
    |       |       | F1    |
    |       |-------|-------|
    |       | F1    | F2    |
    |-------|-------|-------|
    | F1    | F2    | F3    |
     ------- ------- -------
```
Another approach could be to just use two fragment sizes; the largest and the smallest. This approach asymptotes on the single fragment size from #12627.
```
       C0      C1      C2
     -----------------------
RG0 | F0    | F0    | F0    | 24 fragments total
    |       |-------|-------|
    |       | F1    | F1    |
    |       |-------|-------|
    |       | F2    | F2    |
    |-------|-------|-------|
    | F1    | F3    | F3    |
    |       |-------|-------|
    |       | F4    | F4    |
    |       |-------|-------|
    |       | F5    | F5    |
    |=======|=======|=======|
RG1 | F0    | F0    | F0    |
    |       |-------|-------|
    |       | F1    | F1    |
    |       |-------|-------|
    |       | F2    | F2    |
    |-------|-------|-------|
    | F1    | F3    | F3    |
     ------- ------- -------
```
Profiling with a fairly large file (14G uncompressed, 3G snappy compressed, with 112 row groups), it takes about 8 seconds to do the write, `InitPageFragments` takes 11ms, and `RecalculatePageFragments` takes 10ms.  Given the low computational overhead, do you think the alternatives are worth exploring?  I'm not sure if they'll save on code complexity or not.

Reply:
I'm happy with the current approach. I just didn't realize that uniform fragment sizes are required because of row group alignment.

Reply:
Is it worth adding that to the comment and/or renaming the `fragments` array to `rowgroup_fragments` to make this more explicit?

Reply:
Please do, I like both suggestions :)